packages/workspace/src/workspace/config_source.ts

```diff
@@ -39,49 +39,94 @@ class ConfigParseError extends Error {
 
 export const configSource = (
   dirStore: DirectoryStore<string>,
-  configOverrides?: DetailedChange[],
+  configOverrides: DetailedChange[] = [],
 ): ConfigSource => {
   const filename = (name: string): string =>
     (name.endsWith(FILE_EXTENSION) ? name : name.concat(FILE_EXTENSION))
 
   const configOverridesById = _.groupBy(configOverrides, change => change.id.adapter)
 
+  const getConfigWithoutOverrides = async (name: string):
+    Promise<InstanceElement | undefined> => {
+    const naclFile = await dirStore.get(filename(name))
+    if (_.isUndefined(naclFile)) {
+      log.warn('Could not find file %s for configuration %s', filename(name), name)
+      return undefined
+    }
+    const parseResult = await parse(Buffer.from(naclFile.buffer), naclFile.filename)
+    if (!_.isEmpty(parseResult.errors)) {
+      log.error('failed to parse %s due to %o', name, parseResult.errors)
+      throw new ConfigParseError(name)
+    }
+    if (parseResult.elements.length > 1) {
+      log.warn('%s has more than a single element in the config file; returning the first element',
+        name)
+    }
+    const configInstance = parseResult.elements.find(isInstanceElement)
+    if (configInstance === undefined) {
+      log.warn(
+        'failed to find config instance for %s, found the following elements: %s',
+        name,
+        parseResult.elements.map(elem => elem.elemID.getFullName()).join(','),
+      )
+      return undefined
+    }
+    return configInstance
+  }
+
+  const applyConfigOverrides = (conf: InstanceElement): void => {
+    const overridesForInstance = configOverridesById[conf.elemID.adapter] ?? []
+    overridesForInstance.forEach(change => {
+      setPath(conf, change.id, getChangeElement(change))
+    })
+  }
+
+  const setUnsafe = async (name: string, config: InstanceElement): Promise<void> => {
+    await dirStore.set({ filename: filename(name), buffer: await dumpElements([config]) })
+    await dirStore.flush()
+  }
+
+  const validateConfigChanges = (configChanges: DetailedChange[]): void => {
+    const updatedOverriddenIds = configOverrides.filter(
+      overiddeChange => configChanges.some(
+        updateChange => updateChange.id.isParentOf(overiddeChange.id)
+          || overiddeChange.id.isParentOf(updateChange.id)
+          || overiddeChange.id.isEqual(updateChange.id)
+      )
+    )
+
+    if (updatedOverriddenIds.length !== 0) {
+      throw new Error(`cannot update fields that were overriden by the user: ${updatedOverriddenIds.map(change => change.id)}`)
```

Comment:
```suggestion
      throw new Error(`cannot update fields that were overriden by the user: ${updatedOverriddenIds.map(change => change.id.getFullName())}`)
```
I assume the default string representation of the element id won't be very readable